django-stubs: Tighten function signatures with generic Queryset.

PIG208 opened the pull request:
This is a part of #18777.

It uses the generic `QuerySet[_T]` and `ValuesQuerySet[_T, _Row]` supported by `django-stubs` to provide more specific type annotations for `QuerySet`s. This fixes most of the QuerySet-related errors throughout the codebase. A list of affected mypy errors can be found in the collapsable section below (line numbers removed for generating minimal diffs).

#### Fixes:
<details>
<summary>mypy diffs against #18777</summary>

```diff
< analytics/views/activity_common.py error "_QuerySet[Any, Any]" has no attribute "last_visit"  [attr-defined]
< analytics/views/activity_common.py error "_QuerySet[Any, Any]" has no attribute "user_profile"  [attr-defined]
< analytics/views/activity_common.py error "_QuerySet[Any, Any]" has no attribute "user_profile"  [attr-defined]
< analytics/views/activity_common.py error "_QuerySet[Any, Any]" has no attribute "client"  [attr-defined]
< analytics/views/activity_common.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/activity_common.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/activity_common.py error "_QuerySet[Any, Any]" has no attribute "last_visit"  [attr-defined]
14,18d6
< analytics/views/realm_activity.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/realm_activity.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/realm_activity.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/realm_activity.py error "_QuerySet[Any, Any]" has no attribute "user_profile"  [attr-defined]
< analytics/views/stats.py error Missing type parameters for generic type "QuerySet"  [type-arg]
29,34d16
< analytics/views/user_activity.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/user_activity.py error Incompatible return value type (got "_QuerySet[UserActivity, UserActivity]", expected "List[_QuerySet[Any, Any]]")  [return-value]
< analytics/views/user_activity.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/user_activity.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< analytics/views/user_activity.py error "_QuerySet[Any, Any]" has no attribute "client"  [attr-defined]
< analytics/views/user_activity.py error "_QuerySet[Any, Any]" has no attribute "last_visit"  [attr-defined]
59d40
< zerver/actions/create_user.py error Argument "streams" to "bulk_get_subscriber_peer_info" has incompatible type "_QuerySet[Stream, Stream]"; expected "List[Stream]"  [arg-type]
68d48
< zerver/actions/message_edit.py error Argument 2 to "bulk_access_messages" has incompatible type "_QuerySet[Any, Any]"; expected "Sequence[Message]"  [arg-type]
72d51
< zerver/actions/realm_settings.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "Sequence[UserProfile]")  [return-value]
116,117d94
< zerver/lib/cache.py error Incompatible types in assignment (expression has type "_QuerySet[Subscription, int]", variable has type "_QuerySet[Subscription, Subscription]")  [assignment]
< zerver/lib/cache.py error Argument 1 to "display_recipient_cache_key" has incompatible type "Subscription"; expected "int"  [arg-type]
120,122d96
< zerver/lib/cache_helpers.py error Incompatible return value type (got "_QuerySet[RealmCount, int]", expected "List[int]")  [return-value]
< zerver/lib/cache_helpers.py error Incompatible return value type (got "_QuerySet[Stream, Stream]", expected "List[Stream]")  [return-value]
< zerver/lib/cache_helpers.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "List[UserProfile]")  [return-value]
125,127d98
< zerver/lib/digest.py error Argument 1 to "bulk_get_digest_context" has incompatible type "_QuerySet[Any, Any]"; expected "List[UserProfile]"  [arg-type]
< zerver/lib/display_recipient.py error Incompatible return value type (got "_QuerySet[Stream, TypedDict({'name' str, 'id' int})]", expected "List[TinyStreamResult]")  [return-value]
< zerver/lib/email_notifications.py error Incompatible types in assignment (expression has type "List[Message]", variable has type "_QuerySet[Message, Message]")  [assignment]
131,132d101
< zerver/lib/events.py error Need type annotation for "user_messages" (hint "user_messages List[<type>] = ...")  [var-annotated]
< zerver/lib/events.py error Incompatible types in assignment (expression has type "_QuerySet[UserMessage, TypedDict({'message_id' int})]", variable has type "List[Any]")  [assignment]
141,142d109
< zerver/lib/export.py error Incompatible types in assignment (expression has type "Set[int]", variable has type "_QuerySet[Recipient, int]")  [assignment]
< zerver/lib/export.py error Unsupported operand types for & ("Set[int]" and "_QuerySet[Recipient, int]")  [operator]
177,181d143
< zerver/lib/management.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "List[UserProfile]")  [return-value]
< zerver/lib/management.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "List[UserProfile]")  [return-value]
< zerver/lib/management.py error Incompatible types in assignment (expression has type "List[<nothing>]", variable has type "_QuerySet[Any, Any]")  [assignment]
< zerver/lib/management.py error "_QuerySet[Any, Any]" has no attribute "append"  [attr-defined]
< zerver/lib/management.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "List[UserProfile]")  [return-value]
183d144
< zerver/lib/message.py error Incompatible return value type (got "_QuerySet[UserMessage, int]", expected "List[int]")  [return-value]
189,190d149
< zerver/lib/retention.py error Argument 1 to "restore_data_from_archive_by_transactions" has incompatible type "_QuerySet[ArchiveTransaction, ArchiveTransaction]"; expected "List[ArchiveTransaction]"  [arg-type]
< zerver/lib/retention.py error Argument 1 to "restore_data_from_archive_by_transactions" has incompatible type "_QuerySet[ArchiveTransaction, ArchiveTransaction]"; expected "List[ArchiveTransaction]"  [arg-type]
194d152
< zerver/lib/soft_deactivation.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "List[UserProfile]")  [return-value]
209,215d166
< zerver/lib/stream_subscription.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/stream_subscription.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/stream_subscription.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/stream_subscription.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/stream_subscription.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/stream_subscription.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/streams.py error Missing type parameters for generic type "QuerySet"  [type-arg]
217d167
< zerver/lib/subscription_info.py error Missing type parameters for generic type "QuerySet"  [type-arg]
258d207
< zerver/lib/topic.py error Argument 2 to "bulk_access_messages" has incompatible type "_QuerySet[Message, Message]"; expected "Sequence[Message]"  [arg-type]
260,262d208
< zerver/lib/topic.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/topic.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/lib/topic.py error Missing type parameters for generic type "QuerySet"  [type-arg]
268,270d213
< zerver/lib/user_groups.py error Incompatible return value type (got "_QuerySet[UserGroupMembership, int]", expected "List[int]")  [return-value]
< zerver/lib/user_groups.py error Incompatible return value type (got "_QuerySet[UserProfile, UserProfile]", expected "_QuerySet[UserGroup, UserGroup]")  [return-value]
< zerver/lib/user_groups.py error Incompatible types in assignment (expression has type "_QuerySet[UserProfile, int]", variable has type "List[int]")  [assignment]
275,276d217
< zerver/lib/user_topics.py error Argument 1 to "map" has incompatible type "Callable[[Dict[str, Any]], ClauseElement]"; expected "Callable[[TypedDict({'recipient_id' int, 'topic_name' str})], ClauseElement]"  [arg-type]
< zerver/lib/user_topics.py error Incompatible types in assignment (expression has type "List[TypedDict({'recipient_id' int, 'topic_name' str})]", variable has type "_QuerySet[UserTopic, TypedDict({'recipient_id' int, 'topic_name' str})]")  [assignment]
281,283d221
< zerver/lib/users.py error Incompatible default for argument "base_query" (default has type "None", argument has type "_QuerySet[UserProfile, UserProfile]")  [assignment]
< zerver/lib/users.py error Statement is unreachable  [unreachable]
< zerver/lib/users.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "List[UserProfile]")  [return-value]
286,287d223
< zerver/lib/users.py error Argument 1 to "get_custom_profile_field_values" has incompatible type "_QuerySet[CustomProfileFieldValue, CustomProfileFieldValue]"; expected "List[CustomProfileFieldValue]"  [arg-type]
< zerver/lib/users.py error Missing type parameters for generic type "QuerySet"  [type-arg]
292,295d227
< zerver/management/commands/send_custom_email.py error Incompatible types in assignment (expression has type "_QuerySet[Any, Any]", variable has type "List[UserProfile]")  [assignment]
< zerver/management/commands/send_custom_email.py error Incompatible types in assignment (expression has type "_QuerySet[Any, Any]", variable has type "List[UserProfile]")  [assignment]
< zerver/management/commands/send_custom_email.py error Incompatible types in assignment (expression has type "_QuerySet[Any, Any]", variable has type "List[UserProfile]")  [assignment]
< zerver/management/commands/send_custom_email.py error Incompatible types in assignment (expression has type "_QuerySet[Any, Any]", variable has type "List[UserProfile]")  [assignment]
297,298d228
< zerver/management/commands/send_password_reset_email.py error Incompatible types in assignment (expression has type "List[UserProfile]", variable has type "_QuerySet[Any, Any]")  [assignment]
< zerver/management/commands/send_password_reset_email.py error Argument 1 to "send" of "Command" has incompatible type "_QuerySet[Any, Any]"; expected "List[UserProfile]"  [arg-type]
300d229
< zerver/management/commands/soft_deactivate_users.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "List[UserProfile]")  [return-value]
323d251
< zerver/models.py error Argument 1 to "list" has incompatible type "_QuerySet[RealmDomain, TypedDict({'domain' str, 'allow_subdomains' Optional[bool]})]"; expected "Iterable[Dict[str, Union[str, bool]]]"  [arg-type]
332d259
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
334d260
< zerver/models.py error Argument "params" to "extra" of "_QuerySet" has incompatible type "Tuple[Tuple[int, ...]]"; expected "Optional[List[Any]]"  [arg-type]
336a263
> zerver/models.py error Argument "params" to "extra" of "_QuerySet" has incompatible type "Tuple[Tuple[int, ...]]"; expected "Optional[List[Any]]"  [arg-type]
339,340d265
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
342,345d266
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/models.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "Sequence[Stream]")  [return-value]
347d267
< zerver/models.py error Incompatible return value type (got "_QuerySet[Subscription, int]", expected "List[int]")  [return-value]
349d268
< zerver/models.py error Incompatible return value type (got "_QuerySet[Message, Message]", expected "Sequence[Message]")  [return-value]
351d269
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
353c271
< zerver/models.py error Incompatible return value type (got "_QuerySet[DefaultStreamGroup, DefaultStreamGroup]", expected "List[DefaultStreamGroup]")  [return-value]
---
> zerver/models.py error Argument 1 to "add" of "set" has incompatible type "Optional[int]"; expected "int"  [arg-type]
361,366d278
< zerver/models.py error Incompatible return value type (got "_QuerySet[CustomProfileField, CustomProfileField]", expected "List[CustomProfileField]")  [return-value]
< zerver/models.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "Sequence[UserProfile]")  [return-value]
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
< zerver/models.py error Incompatible return value type (got "_QuerySet[Any, Any]", expected "Sequence[UserProfile]")  [return-value]
< zerver/models.py error Missing type parameters for generic type "QuerySet"  [type-arg]
405d316
< zerver/tests/test_custom_profile_data.py error Argument 2 to "try_reorder_realm_custom_profile_fields" has incompatible type "_QuerySet[CustomProfileField, Optional[int]]"; expected "List[int]"  [arg-type]
435,437d345
< zerver/tests/test_import_export.py error Incompatible return value type (got "_QuerySet[Message, Message]", expected "Message")  [return-value]
< zerver/tests/test_import_export.py error "Message" has no attribute "__iter__" (not iterable)  [attr-defined]
< zerver/tests/test_import_export.py error "Message" has no attribute "order_by"  [attr-defined]
440,442d347
< zerver/tests/test_import_export.py error "Message" has no attribute "get"  [attr-defined]
< zerver/tests/test_import_export.py error "Message" has no attribute "get"  [attr-defined]
< zerver/tests/test_import_export.py error "Message" has no attribute "get"  [attr-defined]
445d349
< zerver/tests/test_import_export.py error Argument 1 to "custom_profile_field_values_for" has incompatible type "_QuerySet[CustomProfileField, CustomProfileField]"; expected "List[CustomProfileField]"  [arg-type]
538,541d441
< zerver/tests/test_users.py error Incompatible types in assignment (expression has type "Sequence[UserProfile]", variable has type "_QuerySet[Any, Any]")  [assignment]
< zerver/tests/test_users.py error Incompatible types in assignment (expression has type "Sequence[UserProfile]", variable has type "_QuerySet[Any, Any]")  [assignment]
< zerver/tests/test_users.py error Incompatible types in assignment (expression has type "Sequence[UserProfile]", variable has type "_QuerySet[Any, Any]")  [assignment]
< zerver/tests/test_users.py error Incompatible types in assignment (expression has type "Sequence[UserProfile]", variable has type "_QuerySet[Any, Any]")  [assignment]
551a452,455
> zerver/views/auth.py error Argument 1 to "create_confirmation_link" has incompatible type "PreregistrationUser"; expected "Union[Realm, HasRealmObject, OptionalHasRealmObject]"  [arg-type]
> zerver/views/auth.py note     realm expected "Optional[Realm]", got "ForeignKey[Union[Realm, Combinable, None], Optional[Realm]]"
> zerver/views/auth.py note     realm expected "Realm", got "ForeignKey[Union[Realm, Combinable, None], Optional[Realm]]"
> zerver/views/auth.py note Following member(s) of "PreregistrationUser" have conflicts
604d507
< zerver/views/registration.py error Argument "streams" to "prepare_activation_url" has incompatible type "Optional[_QuerySet[Stream, Stream]]"; expected "Optional[List[Stream]]"  [arg-type]
615d517
< zilencer/management/commands/render_messages.py error Missing type parameters for generic type "QuerySet"  [type-arg]
```
</details>

#### TODOs

- After this, we need to find a different way to correctly type `create_confirmation_link` to accept any type that has `realm` attribute.
- There are still several errors related to the `param` parameter of `.extra` for `QuerySet`s, which needs to be fixed in a follow-up PR.
- For now, we import `_QuerySet` as `ValuesQuerySet` from `django.db.models.query`. But once we get django-stubs integrated, we should import `ValuesQuerySet` from `django_stubs_ext` instead.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

PIG208 commented:
9eb335e4eb got cherry-picked from this branch to #18777, since the overload variants are invalid without django-stubs.

andersk commented on analytics/views/activity_common.py:
```diff
@@ -84,7 +86,9 @@ def remote_installation_stats_link(server_id: int, hostname: str) -> mark_safe:
 
-def get_user_activity_summary(records: List[QuerySet]) -> Dict[str, Any]:
+def get_user_activity_summary(
+    records: Union[Sequence[UserActivity], QuerySet[UserActivity]]
```
Could be `records: Collection[UserActivity]`, right?

andersk commented on zerver/lib/digest.py:
```diff
@@ -276,7 +277,9 @@ def get_slim_stream_map(stream_ids: Set[int]) -> Dict[int, Stream]:
 
-def bulk_get_digest_context(users: List[UserProfile], cutoff: float) -> Dict[int, Dict[str, Any]]:
+def bulk_get_digest_context(
+    users: Union[Sequence[UserProfile], QuerySet[UserProfile]], cutoff: float
```
Could be `users: Collection[UserProfile]` if we changed `realm = users[0].realm` to `realm = next(iter(users)).realm`.

andersk commented on zerver/lib/message.py:
```diff
@@ -835,7 +847,7 @@ def has_message_access(
 
 def bulk_access_messages(
-    user_profile: UserProfile, messages: Sequence[Message], *, stream: Optional[Stream] = None
+    user_profile: UserProfile, messages: Iterable[Message], *, stream: Optional[Stream] = None
```
Prefer `messages: Collection[Message]` here because we iterate it twice. A call like `bulk_access_messages(user, (message for message in messages))` should be rejected, because although it won’t throw a type error, the generator will be unexpectedly empty for the second iteration.

andersk commented on zerver/lib/stream_subscription.py:
```diff
@@ -164,7 +164,7 @@ def get_user_ids_for_streams(stream_ids: Set[int]) -> Dict[int, Set[int]]:
 def bulk_get_subscriber_peer_info(
     realm: Realm,
-    streams: List[Stream],
+    streams: Iterable[Stream],
```
`Collection[Streams]` because we iterate it twice.

andersk commented on zerver/lib/management.py:
```diff
@@ -116,7 +117,7 @@ def get_users(
         is_bot: Optional[bool] = None,
         include_deactivated: bool = False,
-    ) -> Iterable[UserProfile]:
+    ) -> Union[List[UserProfile], QuerySet[UserProfile]]:
```
This should go in the first commit. `Collection[UserProfile]` might work.

PIG208 commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -743,7 +743,8 @@ def test_list_order(self) -> None:
             .values_list("order", flat=True)
         )
-        try_reorder_realm_custom_profile_fields(realm, order)
+        assert all([isinstance(item, int) for item in order])
```
This is the exact error that I got without the cast: `Argument 2 to "try_reorder_realm_custom_profile_fields" has incompatible type "_QuerySet[CustomProfileField, Optional[int]]"; expected "Iterable[int]"`. Possibliy a bug in the plugin.

andersk replied:
Looks like typeddjango/django-stubs#444.

PIG208 replied:
Assuming that this will get fixed in the future, I took out the commit.

PIG208 commented on zerver/models.py:
```diff
@@ -1075,7 +1075,12 @@ class EmailContainsPlusError(Exception):
-def get_realm_domains(realm: Realm) -> List[Dict[str, Union[str, bool]]]:
+class RealmDomainDict(TypedDict):
+    domain: str
+    allow_subdomains: Optional[bool]
```
Similarily, we have `Argument 1 to "list" has incompatible type "_QuerySet[RealmDomain, TypedDict({'domain': str, 'allow_subdomains': Optional[bool]})]"; expected "Iterable[RealmDomainDict]"`. Will need to fix this upstream.

PIG208 commented:
`allow_subdomains` is made non-optional. There is a similar false-positive at [here](https://github.com/zulip/zulip/blob/7c992422f01e2091b50f4d0db997f19619b36cf0/zerver/management/commands/send_custom_email.py#L95-L97).

I rebased the PR, refactored [get_ids](https://github.com/pig208/zulip/blob/queryset-filterbranch/zerver/lib/export.py#L1201-L1202) in 0f217477e7925be73bb31171ff918582f3ecb2d3, and added a new commit 268a837e10c56b80cc2d16fa22c637f125bf5ed7. This is otherwise unchanged.

andersk commented on zerver/management/commands/sync_ldap_user_data.py:
```diff
@@ -18,7 +18,7 @@
 @transaction.atomic
 def sync_ldap_user_data(
-    user_profiles: List[UserProfile], deactivation_protection: bool = True
+    user_profiles: Iterable[UserProfile], deactivation_protection: bool = True
```
This is iterated on both line 25 and line 27, so should be `Collection[UserProfile]`.

andersk commented on zerver/lib/management.py:
```diff
@@ -140,10 +140,10 @@ def get_users(
-        user_profiles = []
+        user_profiles_list = []
         for email in emails:
-            user_profiles.append(self.get_user(email, realm))
-        return user_profiles
+            user_profiles_list.append(self.get_user(email, realm))
+        return user_profiles_list
```
Could simplify to `return [self.get_user(email, realm) for email in emails]`.

andersk commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -743,7 +743,8 @@ def test_list_order(self) -> None:
             .values_list("order", flat=True)
         )
-        try_reorder_realm_custom_profile_fields(realm, order)
+        assert all(isinstance(item, int) for item in order)
+        try_reorder_realm_custom_profile_fields(realm, cast(Iterable[int], order))
```
This should have a comment pointing at https://github.com/typeddjango/django-stubs/issues/444, and/or we should encourage upstream to fix it, since it looks pretty simple.

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -121,11 +121,9 @@ def exclude_topic_mutes(
+        "topic_name",
     )
 
     if not rows:
```
I'm not sure, but I think removing the `list()` wrapper here might add an extra database query -- I think `if not rows` is going to do a `.exists()` query, and then we'll do a second query when accessing `rows` later.

PIG208 replied:
I believe that the result is [cached](https://github.com/django/django/blob/6f80050496780e2f8a65e2a4374dd6375c9fbfa5/django/db/models/query.py#L2070-L2076) once either `__iter__` and `__bool__` is called.
